samples/ha-demo/platform/kube/ha-demo.yaml

```diff
@@ -0,0 +1,265 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: ha-demo
+  labels:
+    istio-injection: enabled
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: client
+  labels:
+    istio-injection: disabled
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: front-envoy
+  namespace: ha-demo
+data:
+  front-envoy.yaml: |
```

Comment:
This is really confusing to me. Do we have no istio-proxy and are just directly using Envoy? Or do we have envoy with an envoy sidecar?

Reply:
Agreed - this is a bit confusing. We are using a simple envoy for the workload and the sidecar adds another envoy. We actually needed to add a parameter to allow two envoys on one machine. The `yaml` in this file has nothing to do with the sidecar. It is only for the demo app.

Reply:
Why is this done in such a complicated way? You could just as easily demonstrate HA by continuously restarting bookinfo services, couldn't you? That way, you would use Istio, instead of Envoy, configuration yaml and it would be a proper "Istio HA" example. This looks more like an Envoy HA experiment than Istio.

Reply:
> Why is this done in such a complicated way? 

Surprise, that was our first approach :) There are several reasons, why we decided to do it this way. So we wanted to check if this is also of interest for the community.

* The plain bookinfo app is not ha ready (replica count, grace period, .. )
* The flask web server is is explicitely documented to not use in productive environments. So using it in a ha demo sounds strange, right?
* You need a sidecar for the bookinfo to work, right? That would mean that we have to restart the client when we update istio (to get the new sidecar) and we are of course not able to see if there are errors during the switch
* Testing the bookinfo via ingress is not an option either either, since  how to update the ingress in an ha manner in still unclear to me.

Reply:
I think you missed the main point. IMHO, you need something to be ha first - without istio. If this is rock solid, you can use istio and check if it is still working. In our first approach, we used the bookinfo app and registered regularly 503 from the microservices. But then it was hard to judge if the error is coming from istio, the bookinfo app or kubernetes.

Reply:
@c0d1ngm0nk3y I'm not really following your explanation. It's unclear to me exactly what you are trying to test. Let me try to explain what I think the sample is doing.

The sample has 3 services:
1. `news`
2. `account`
3. `account-v2`

and 2 deployments:
1. `news` (implements `news` service)
2. `account-v2` (implements `account-v2`)

You seem to be using Istio to redirect calls to the `account` service, from `news`, to the `accounts-v2` service instead. That's sort of a weird way to use Istio, normally you would be redirecting to a specific subset (version) of the `account` service - you would not even have an `account-v2` service, just an `account-v2` deployment. Is there something special about this contrived approach that is needed to demonstrate HA??? The other thing that seems odd is the way that you've implemented the `news` service as an envoy calling another envoy (sidecar). The first envoy has hard-coded envoy conifg path match. Why wouldn't that be configured with a VirtualService instead? Why is the workload an envoy in the first place? This is all very confusing. If there are reasons this needs to be so complicated, it needs to be explained in the example/blog document.

Next you repeatedly call the news service from a non-Istio-enabled pod (sleep sample), while continuously restarting the news and account-v2 pods.

So it looks to me like you are trying to prove that calls from outside the mesh (i.e., curls from the external sleep sample) don't fail when the two istio-enabled pods are restarted. You seem, however, to keep talking about restarting Istio, which your sample is not doing. Demonstrating HA while upgrading/restarting Istio would be a good sample, but that's not what you're doing, as far as I can tell. You're just restarting istio-enabled pods, which happens all the time, if you enable autoscaling, for example. Is there something more subtle here?

Anyway, the bottom line is that I don't think users will understand what this sample is doing unless they spend a fair bit of time digging though it, like I did, because the article does not explain enough, and I still feel that unless I'm missing something it is unnecessarily complicated and not using Istio the way an actual application (that wants to be HA) would. You also need to be very specific and clear in the article about what exactly the sample is doing, why it is so non-conventional, and what kind of HA it is proving.

Reply:
@frankbu Fair points. I think it is clear that this example can/should be improved to make it more crisp. Also you are right, it does not change the control plane yet. But I am personally not a fan of `big bang developments`. So it was important to first get some feedback. I also don't think that this news/article topic is perfect, but renaming is not a problem.

The important part was to have such kind of application/demo. Which can then later be used to show more, e.g. updating istio without downtime. IMHO, the bookinfo is in no way suitable for such tests.

But I have understood the point that this is to envoy-ish, so it might make no sense to work on better naming/documentation. I will close those PR and we will use this for our tests/validations of the installer/update.

Thanks for the review.